### ISSUE TITLE
How many GPUs are used in training?

### ISSUE THREAD

**Issue by xiaohangt** [autonomousvision/mdpo#5]
Hi, thanks for the great work, I'm recently working on dLLMs and really appreciate your effort on optimizing the codebase!

I tried to run training with the default hyperparams with 4 A100 but it is very slow. May I ask how many GPUs are used to train 3500 samples within ~30hrs as noted in the README?

Thanks!

**Comment by Cli212** [comment]
Hi, thanks for the comments. In our experiments we use 8 H100s. 

May I ask which part is slow? The inference part that filters out the Answer Backslide samples or the training part? I am considering uploading the list of Answer Backslide samples idx so that you don't need to run the inference again to filter out the same samples I already have. 

Note that the progress bar in training actually does not reflect the correct amount of hours needed if you turn on --incremental_training. Because under this setting only Answer Backslide samples will be used which is ~10% of the data. So you should also expect that 10% of the shown training hours in the progress bar are needed. I will fix this quickly.

**Comment by xiaohangt** [comment]
Thanks for the quick response! Yes I turned on --incremental _training.

I'm looking at the weight and bias, which showed that 6 global steps took over 5 hrs. This cost is very different from d1 (diffu-GRPO) --- 5K steps taking around 24hrs. I'm thinking is it because of large `gradient_accumulation_steps=16` and `max_completion_length==512`. The usage of intermediate responses and RCR shouldn't cause such a huge slow down.

**Comment by xiaohangt** [comment]
Sorry I just realize the first 200 minutes are used to do `_incremental_data_selection`. And I think this is one-off. By considering that, the training speed seems close to normal.

**Comment by Cli212** [comment]
I see. Though I think the training speed will still be slower than diffu-GRPO as we use larger gradient_accumulation_steps. Also because each intermediate step is actually a data sample (creates gradients) in our training, the actual batch size is batch_size * N_gpus * gradient_accumulation_steps * **sample_train_steps** (number of intermediate steps we use for gradient update). So it should still be sample_train_steps * (our_gradient_accumulation_steps / d1_gradient_accumulation_steps) fewer global steps given the same amount of time. However, I would expect that with around 30 global updates you will already see the model improves on validation set as our method is sample-efficient.

**Comment by xiaohangt** [comment]
Thanks for the detailed explanations, very helpful!

**Comment by xiaohangt** [comment]
Hi! Sorry to bother again. Currently facing some trouble reproduce the results in the paper. Would be great to have your suggestions!
- In README, `remask_strategy` is set to random and it is claimed that this is consistent with the paper's setting. Is it a typo? Besides, `overtime_confidence` is missing. I tried to change it to `low_confidence` and overtime, the eval results have some improvement.
- I enable `overtime_confidence` and `incremental_data_selection`, but only got 36.4% on MATH500 with 1 epoch training, and it's already the best among all my runnings. I didn't change any other things. Is there any thing I'm missing? Do you think I should do more epochs?
- If there is any other potential direction I can do to improve the accuracy to match the number?

**Comment by Cli212** [comment]
Hi, thanks for the comments. 

- Yes, it's a typo to set remask_strategy as random, I changed the README to have it as low_confidence now. BTW, for creating rollouts we didn't use overtime_confidence as well. We don't know how it would affect the training performance.
- This marginal improvement is natural if you only use 3500 samples with incremental_training because then it leads to only ~400 filtered Answer Backslide (AB) training data (4 total gradient update steps if you use gradient accumulation 128)? Our training is applied on the AB data samples filtered from 40000 samples. I just uploaded the samples to the folder ab_samples and you could easily use them by --ab_path ab_samples/ab_from_40k_epoch_1.csv. This is to save you some time to avoid running the inference before the first epoch. And don't forget to set --num_train_samples 40000 even though it doesn't consume all the data, just for the huggingface trainer to allocate enough update steps.
- With the above setting, you should already be able to reproduce our experiment results, let me know if there is anything unclear.

**Comment by xiaohangt** [comment]
Thanks a lot for the guidance! I manage to obtain 43.4% on my machines, which is close enough. Thanks for sharing the data, these are really helpful!
